### PR TITLE
When a signed request fails, fire an action so that sites can log the details of the failure

### PR DESCRIPTION
**Pull request by mdawaffe**
#### Changes proposed in this Pull Request:
Adds a `jetpack_verify_signature_error` action that is fired when an XML-RPC or REST API request's signature is invalid.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
New feature: ability to do better error reporting.

#### Testing instructions:
WIP

#### Proposed changelog entry for your changes:
None required?

### CODE REVIEW

File: class.jetpack.php [hunk 1]
```diff
@@ -5395,6 +5422,18 @@ function wp_rest_authenticate( $user ) {
...
+	protected function send_signature_error_header( $error ) {
+		$error_data = $error->get_error_data();
+		if ( ! isset( $error_data['signature_details'] ) ) {
+			return;
+		}
+
+		header( sprintf(
+			'X-Jetpack-Signature-Error: %s',
+			base64_encode( json_encode( $error_data['signature_details'] ) )
+		) );
```
Comment (eliorivero) [inline class.jetpack.php]: Can you simplify the logic here so the function doesn't return void on one case and nothing on the other?
```
		if ( isset( $error_data['signature_details'] ) ) {
			header( sprintf(
				'X-Jetpack-Signature-Error: %s',
				base64_encode( json_encode( $error_data['signature_details'] ) )
			) );
		}
```

Comment (gravityrail) [reply class.jetpack.php]: Why does it matter what this function returns? We are not checking the output anywhere.

Nested if's don't really simplify the logic IMO, they just move it around.

**Comment by gravityrail** [comment]
I dig the approach, haven't tested interactively myself but seems like a good idea.

I was initially skeptical about always returning the header if the signature is invalid, since some endpoints don't require a valid signature and therefore it doesn't matter, but I'm coming around to this way of doing things.

File: class.jetpack.php (hunk 1 above)
Comment (mdawaffe) [reply class.jetpack.php]: There is no practical difference between `return;` and not calling `return`. Both return `null`:

```php
function return_nothing() {
  return;
}

function do_not_return() {
   // nothing
}

function return_null() {
  return null;
}

var_dump( return_nothing(), do_not_return(), return_null() ); // NULL, NULL, NULL
```

The only difference between `void` and `null` occurs when using `void` as a return type hint.

```php
function this_works() : void {
  return; // fine
}

function this_also_works() : void {
   // nothing - also fine
}

function this_breaks() : void {
  return null; // not fine
}
```

The last function will cause a parse error.

So PHP knows the difference, but calling code can't. (Well - there might be a way to tell the difference with reflection, but that's silly :))

In the case of this `send_signature_error_header()` method, if we were to include a `void` return type hint, everything would be fine, so my claim is that this falls into the "there is no practical difference" category :)

I think an aesthetic/code style argument could be made, but I (subjectively) disagree ;) I prefer the "bail as early as possible and indent as little as possible" argument in this case.

**Comment by mdawaffe** [comment]
I've thought about this some more, and I don't like the idea of exposing the exact error state; I'm worried it leaks too much information. We may want to do this eventually, but I suggest being more conservative for now: Instead of always responding with exactly what failed, just fire an action.

We can add logging or use the header on Atomic sites to gather some data in a more controlled fashion than we could if we turned this on everywhere.

File: class.jetpack.php (hunk 1 above)
Comment (mdawaffe) [reply class.jetpack.php]: (Moot now because the code is gone ;))

**Review by gravityrail — DISMISSED** [review]
This looks good to me. Merge when ready.

**Review by zinigor — DISMISSED** [review]
LGTM, removing mixed signals, like that giant DO NOT MERGE label.

**Comment by mdawaffe** [comment]
I just added some more debugging details: the provided signature and the expected signature (when we know it).

Prior to this latest commit, I was reasonably sure all this debugging information could be made public. I switched from the original method of always sending the debugging information back in the response to the new method of just firing an action so that the debugging can be logged someplace custom out of caution: I believe the debugging data is safe, but if it does expose a vulnerability, better that we can control it without having to ship a new version of the plugin.

After this latest commit, I no longer believe the debugging information is safe to expose publicly. Giving direct feedback about what signature was expected in general seems very sketchy to me. For this specific case, it would be disastrous without a rewrite of how we handle things like duplicate requests.

If we ever do want to make this data publicly available, we'll need to be careful.

**Comment by mdawaffe** [comment]
I'm still thinking a couple things through :) Will be ready for review shortly.

**Comment by mdawaffe** [comment]
906206aae89f00efd2922f6d4f3eb57771cee068 makes the debugging info safer (not safe :)) to expose publicly. The request duplication detection code (the nonce stuff) is now run before we add the expected signature to the debugging details.

File: class.jetpack.php [hunk 2]
```diff
@@ -5165,23 +5164,55 @@ public function reset_saved_auth_state() {
...
+	 */
 	function verify_xml_rpc_signature() {
-		if ( $this->xmlrpc_verification ) {
-			return $this->xmlrpc_verification;
+		if ( is_null( $this->xmlrpc_verification ) ) {
+			$this->xmlrpc_verification = $this->internal_verify_xml_rpc_signature();
+
+			if ( is_wp_error( $this->xmlrpc_verification ) ) {
+				// Action for logging signature verification errors. This data is sensitive.
+				do_action( 'jetpack_verify_signature_error', $this->xmlrpc_verification );
```
Comment (jeherve) [inline class.jetpack.php]: Could you add a docblock for that action, to make our parser happy?

**Review by jeherve — DISMISSED** [review]

**Comment by gravityrail** [comment]
I added that missing docblock

**Review by zinigor — DISMISSED** [review]
Looks like this one is going to be green: I have merged the conflicting changes back to this branch. Feel free to merge this one!

**Review by jeherve — APPROVED** [review]
This works for me. Merging.

**Pull request merged**